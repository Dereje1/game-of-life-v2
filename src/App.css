canvas {
  outline: none;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0); /* mobile webkit */
}
