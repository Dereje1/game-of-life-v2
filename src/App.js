import React, { Component } from 'react';
import { ControlTop, ControlBottom } from './components/Controls';
import { getActiveCells, getNewLiveCells } from './components/utils'

class App extends Component {

    constructor(props) {
        super(props);
        this.state = {
            refresh: false,
            cellSize: 20,
            refreshRate: 140,
            generations: 0,
            empty: false
        };
        this.Canvas = React.createRef();
    }

    componentDidMount() {
        this.loadCanvas()
    }

    onRefresh = () => {
        this.setState({ refresh: true }, this.refreshCells)
    }

    onReset = () => {
        this.setState({ refresh: false, generations: 0 }, this.setGrid)
    }

    onClear = () => {
        this.setState({ refresh: false, generations: 0, empty: true }, this.setGrid)
    }

    handleCellSize = ({ target: { value } }) => {
        this.setState({ refresh: false, cellSize: value }, this.loadCanvas)
    }

    onRefreshRate = ({ target: { value } }) => {
        clearTimeout(this.timeoutId);
        this.setState({ refreshRate: value }, this.updateCells)
    }

    refreshCells = () => {
        const { cells: oldCells, cellSize, canvasWidth: width, canvasHeight: height, generations } = this.state;
        const newLiveCells = getNewLiveCells(oldCells, cellSize, width, height);
        const keys = Object.keys(oldCells);
        const cells = {};
        for (const cell of keys) {
            if (newLiveCells.includes(cell)) {
                cells[cell] = {
                    isAlive: true
                }
            } else {
                cells[cell] = {
                    isAlive: false
                }
            }
        }
        this.setState({ cells, generations: generations + 1 }, this.updateCells)
    }

    updateCells = () => {
        const { cells, canvasWidth, canvasHeight, refresh, cellSize, refreshRate } = this.state;
        const context = this.canvasContext
        const activeCells = getActiveCells(cells);
        context.clearRect(0, 0, canvasWidth, canvasHeight);
        activeCells.forEach(cell => {
            const [x, y] = cell.split('-');
            context.fillStyle = "#ff0000";
            context.fillRect(x, y, cellSize, cellSize)
        })
        this.refreshGrid();
        if (refresh && activeCells.length) {
            clearTimeout(this.timeoutId);
            this.timeoutId = setTimeout(this.refreshCells, refreshRate);
        }
    }

    handleCanvasClick = ({ clientX, clientY }) => {
        const { cells, cellSize } = this.state
        const Canvas = this.Canvas.current;
        const { left, top } = Canvas.getBoundingClientRect();
        const trueX = clientX - left;
        const trueY = clientY - top;
        const xFloor = Math.floor(trueX / cellSize) * cellSize
        const yFloor = Math.floor(trueY / cellSize) * cellSize
        const key = `${xFloor}-${yFloor}`
        this.setState({
            cells: {
                ...cells,
                [key]: {
                    isAlive: !cells[key].isAlive
                }
            }
        }, this.updateCells)
    }

    setGrid = () => {
        const { canvasWidth, canvasHeight, cellSize, empty } = this.state;
        const cells = {}
        for (let x = 0; x < canvasWidth; x += cellSize) {
            for (let y = 0; y < canvasHeight; y += cellSize) {
                cells[`${x}-${y}`] = {
                    isAlive: empty ? false : Math.random() < 0.5
                }
            }
        }
        this.setState({ cells, empty: false }, this.updateCells)
    }

    refreshGrid = () => {
        const context = this.canvasContext
        const { canvasWidth, canvasHeight, cellSize } = this.state;
        for (let x = 0; x <= canvasWidth; x += cellSize) {
            context.beginPath();
            context.moveTo(x, 0);
            context.lineTo(x, canvasHeight);
            context.strokeStyle = "#3b3b3b";
            context.stroke();
        }
        for (let y = 0; y <= canvasHeight; y += cellSize) {
            context.beginPath();
            context.moveTo(0, y);
            context.lineTo(canvasWidth, y);
            context.strokeStyle = "#3b3b3b";
            context.stroke();
        }
    }

    loadCanvas = () => {
        const Canvas = this.Canvas.current;
        const { innerWidth, innerHeight } = window;
        const { cellSize } = this.state
        const cellWidth = Math.floor(innerWidth / cellSize);
        const cellHeight = Math.floor((innerHeight * 0.8) / cellSize);
        const canvasWidth = cellWidth * cellSize;
        const canvasHeight = cellHeight * cellSize

        if (Canvas) {
            Canvas.style.backgroundColor = 'black';
            this.canvasContext = Canvas.getContext('2d');
            this.setState({
                canvasWidth,
                canvasHeight,
                canvasTop: (innerHeight - canvasHeight) / 2,
                canvasLeft: (innerWidth - canvasWidth) / 2
            }, this.setGrid);
        }
    }

    render() {
        const {
            canvasWidth, canvasHeight,
            canvasTop, canvasLeft,
            cellSize, refreshRate,
            refresh, generations
        } = this.state;
        return (
            <>
                <>
                    <ControlTop
                        width={canvasWidth}
                        height={window.innerHeight * 0.1}
                        isRefreshing={refresh}
                        onRefresh={this.onRefresh}
                        onPause={() => this.setState({ refresh: false })}
                        onReset={this.onReset}
                        onClear={this.onClear}
                        generations={generations} />


                    <ControlBottom
                        width={canvasWidth}
                        height={window.innerHeight * 0.1}
                        cellSize={cellSize}
                        refreshRate={refreshRate}
                        handleCellSize={this.handleCellSize}
                        onRefreshRate={this.onRefreshRate} />
                </>
                <div style={{ paddingLeft: canvasLeft }}>
                    <canvas
                        ref={this.Canvas}
                        width={canvasWidth}
                        height={canvasHeight}
                        tabIndex="0"
                        onClick={this.handleCanvasClick} />
                </div>
            </>
        )
    }
}

export default App