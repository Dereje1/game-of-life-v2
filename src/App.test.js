import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import App from "./App";

jest.useFakeTimers();

const initialProps = {
  refresh: false,
  cellSize: 15,
  refreshRate: 125,
  generations: 0,
  showGrid: true,
  showPatternDialog: false,
  patternName: "none",
  refreshVal: 3,
  showSettingsDialog: false,
  cells: [],
  generationsPerSecond: 0,
  canvasWidth: 0, // width of the canvas set after CDM
  canvasHeight: 0, // height of the canvas set after CDM
  canvasLeft: 0, // left padding of the canvas set after CDM
  isMaxElemets: false, // true if total cells exceeds allotted
  metricTimeStamp: Date.now(), // timestamp used for metrics/perf display
  colors: {
    canvasBackGround: "black",
    liveCell: "yellow",
    grid: "#3b3b3b"
  },
  showColorPicker: false,
  selectedColorType: "canvasBackGround"
};

let useRefSpy;
const clearRect = jest.fn();
const fillRect = jest.fn();
const focus = jest.fn();

beforeEach(() => {
  useRefSpy = jest.spyOn(React, "createRef").mockImplementation(() => ({
    current: {
      style: {},
      getContext: () => ({
        clearRect,
        fillRect,
        beginPath: jest.fn(),
        moveTo: jest.fn(),
        lineTo: jest.fn(),
        stroke: jest.fn()
      }),
      getBoundingClientRect: () => ({ left: 1, top: 1 }),
      focus
    }
  }));
  global.innerWidth = 90;
  global.innerHeight = 120;
});
afterEach(() => {
  useRefSpy.mockClear();
  clearRect.mockClear();
  fillRect.mockClear();
  focus.mockClear();
  global.innerWidth = 0;
  global.innerHeight = 0;
});

test("renders", () => {
  const wrapper = shallow(<App {...initialProps} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test("will set the size of the canvas and cells", () => {
  const wrapper = shallow(<App {...initialProps} />);
  expect(wrapper.state().cells).toStrictEqual([]);
  expect(wrapper.state().canvasWidth).toBe(90);
  expect(wrapper.state().canvasHeight).toBe(90);
  expect(wrapper.state().cellSize).toBe(15);
});

test("will limit the canvas size if # elements exceeds allotted", () => {
  global.innerWidth = 3500;
  global.innerHeight = 3500;
  const wrapper = shallow(<App {...initialProps} />);
  expect(wrapper.state().cells).toStrictEqual([]);
  expect(wrapper.state().canvasWidth).toBe(1440);
  expect(wrapper.state().canvasHeight).toBe(655);
  expect(wrapper.state().cellSize).toBe(5);
});

test("will draw active cells on canvas", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ cells: [7] });
  wrapper.instance().updateCells();
  //first call for clearRect on CDM, second one on updateCells
  expect(clearRect).toHaveBeenCalledTimes(2);
  expect(clearRect).toHaveBeenCalledWith(0, 0, 90, 90);
  expect(fillRect).toHaveBeenCalledTimes(1);
  expect(fillRect.mock.lastCall).toEqual([15, 15, 15, 15]);
});

test("will refresh active cells", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.instance().handlePattern();
  // blinker oscillator
  const liveCells = [6, 7, 8];
  wrapper.setState({
    cells: liveCells,
    refresh: true,
    generations: 4,
    patternName: "xyz"
  });
  wrapper.instance().updateCells();
  jest.advanceTimersByTime(140);
  const refreshedCells = [7, 1, 13];
  expect(wrapper.state().cells).toStrictEqual(refreshedCells);
  expect(wrapper.state().refresh).toBe(true);
  expect(wrapper.state().patternName).toBe("xyz");
  expect(wrapper.state().generations).toBe(5);
});

test("will stop refresh of active cells if no live cells found", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.instance().handlePattern();
  const liveCells = [7];
  wrapper.setState({
    cells: liveCells,
    refresh: true,
    generations: 4,
    patternName: "xyz"
  });
  wrapper.instance().updateCells();
  jest.advanceTimersByTime(140);
  expect(wrapper.state().cells).toStrictEqual([]);
  expect(wrapper.state().refresh).toBe(false);
  expect(wrapper.state().patternName).toBe("none");
  expect(wrapper.state().generations).toBe(4);
});

test("will make cells active on click", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.instance().handlePattern();
  expect(wrapper.state().cells.includes(7)).toBe(false);
  wrapper.instance().handleCanvasClick({ clientX: 18, clientY: 18, preventDefault: jest.fn() });
  expect(wrapper.state().cells.includes(7)).toBe(true);
});

test("will make cells in-active on click", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ cells: [7, 8] });
  wrapper.instance().handleCanvasClick({ clientX: 18, clientY: 18, preventDefault: jest.fn() });
  expect(wrapper.state().cells).toEqual([8]);
});

test("will handle refreshing cells on play", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ refresh: false, cells: [6, 7, 8] });
  const controlTop = wrapper.find("ControlTop");
  controlTop.props().handleRefresh();
  expect(wrapper.state().refresh).toBe(true);
});

test("will handle stopping cell refresh on pause", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ refresh: true });
  const controlTop = wrapper.find("ControlTop");
  controlTop.props().handlePause();
  expect(wrapper.state().refresh).toBe(false);
});

test("will handle showing the pattern dialog", () => {
  const wrapper = shallow(<App {...initialProps} />);
  const controlTop = wrapper.find("SettingsDialog");
  controlTop.props().handlePatternDialog();
  expect(wrapper.state().showPatternDialog).toBe(true);
});

test("will handle clearing all live cells in the grid", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ refresh: true, generations: 10 });
  const controlTop = wrapper.find("ControlTop");
  controlTop.props().handleClear();
  expect(wrapper.state().refresh).toBe(false);
  expect(wrapper.state().generations).toBe(0);
});

test("will handle adjusting the cellSize", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ refresh: true, generations: 10 });
  expect(wrapper.state().cellSize).toBe(15);
  const controlBottom = wrapper.find("ControlBottom");
  controlBottom.props().handleCellSize({ target: { value: 32 } });
  expect(wrapper.state().cellSize).toBe(32);
});

test("will handle adjusting the refresh rate", () => {
  const wrapper = shallow(<App {...initialProps} />);
  expect(wrapper.state().refreshRate).toBe(125);
  const controlBottom = wrapper.find("ControlBottom");
  controlBottom.props().handleRefreshRate({ target: { value: 2 } });
  expect(wrapper.state().refreshRate).toBe(250);
});

test("will toggle the grid", () => {
  const wrapper = shallow(<App {...initialProps} />);
  expect(wrapper.state().showGrid).toBe(true);
  const controlTop = wrapper.find("SettingsDialog");
  controlTop.props().handleGrid();
  expect(wrapper.state().showGrid).toBe(false);
});

test("will cancel the pattern dialog", () => {
  const wrapper = shallow(<App {...initialProps} />);
  const controlTop = wrapper.find("SettingsDialog");
  controlTop.props().handlePatternDialog();
  expect(wrapper.state().showPatternDialog).toBe(true);
  const patternsDialog = wrapper.find("PatternsDialog");
  patternsDialog.props().handleCancel();
  expect(wrapper.state().showPatternDialog).toBe(false);
});

test("will update the selected pattern", () => {
  const wrapper = shallow(<App {...initialProps} />);
  expect(wrapper.state().patternName).toBe("none");
  const patternsDialog = wrapper.find("PatternsDialog");
  patternsDialog.props().handlePatternChange({ target: { value: "new pattern" } });
  expect(wrapper.state().patternName).toBe("new pattern");
});

test("will hide the settings dialog when a new pattern is selected", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ showSettingsDialog: true });
  const patternsDialog = wrapper.find("PatternsDialog");
  patternsDialog.props().handleOk();
  expect(wrapper.state().showSettingsDialog).toBe(false);
});

test("will handle focus on entering of the pattern dialog", () => {
  const wrapper = shallow(<App {...initialProps} />);
  const patternsDialog = wrapper.find("PatternsDialog");
  patternsDialog.props().handleEntering();
  expect(focus).toHaveBeenCalledTimes(1);
});

test("will handle not showing the settings dialog", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ showSettingsDialog: true });
  const settingsDialog = wrapper.find("SettingsDialog");
  settingsDialog.props().handleClose();
  expect(wrapper.state().showSettingsDialog).toBe(false);
});

test("will handle showing the color picker and hiding the settings dialog", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ showSettingsDialog: true });
  const settingsDialog = wrapper.find("SettingsDialog");
  settingsDialog.props().handleColorPicker();
  expect(wrapper.state().showColorPicker).toBe(true);
  expect(wrapper.state().showSettingsDialog).toBe(false);
});

test("will handle not showing the colors dialog", () => {
  const wrapper = shallow(<App {...initialProps} />);
  wrapper.setState({ showColorPicker: true });
  const colorsDialog = wrapper.find("ColorsDialog");
  colorsDialog.props().closeColorPicker();
  expect(wrapper.state().showColorPicker).toBe(false);
});

test("will handle setting the change color type from the colors dialog", () => {
  const wrapper = shallow(<App {...initialProps} />);
  const colorsDialog = wrapper.find("ColorsDialog");
  colorsDialog.props().updateColorChangeType("liveCell");
  expect(wrapper.state().selectedColorType).toBe("liveCell");
});

test("will handle setting the selected color from the color picker", () => {
  const wrapper = shallow(<App {...initialProps} />);
  expect(wrapper.state().colors.canvasBackGround).toBe("black");
  const colorsDialog = wrapper.find("ColorsDialog");
  colorsDialog.props().handleColorChange({ hex: "FFFFFF" });
  expect(wrapper.state().colors.canvasBackGround).toBe("FFFFFF");
});
