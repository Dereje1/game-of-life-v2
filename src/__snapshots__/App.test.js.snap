// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Fragment>
  <ControlTop
    generations={0}
    handleClear={[Function]}
    handleGrid={[Function]}
    handlePause={[Function]}
    handleRefresh={[Function]}
    handleReset={[Function]}
    height={12}
    isRefreshing={false}
    showGrid={true}
  />
  <ControlBottom
    cellSize={15}
    handleCellSize={[Function]}
    handleRefreshRate={[Function]}
    height={12}
    refreshRate={140}
  />
  <div
    style={
      Object {
        "paddingLeft": 0,
      }
    }
  >
    <canvas
      height={90}
      onClick={[Function]}
      tabIndex="0"
      width={90}
    />
  </div>
  <PatternsDialog
    handleCancel={[Function]}
    handleChange={[Function]}
    handleOk={[Function]}
    open={false}
    value="Random"
  />
</Fragment>
`;
